File: leecode/leetcode-cn/1004.最大连续1的个数III.java
/*
 * @Author: your name
 * @Date: 2022-03-29 21:15:27
 * @LastEditTime: 2022-03-29 21:15:27
 * @LastEditors: Please set LastEditors
 * @Description: 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 * @FilePath: \practice\leecode\leetcode-cn\1004.最大连续1的个数III.java
 */


// 本题和「2024. 考试的最大困扰度」完全一致。
// @ref 官方题解
// 滑动窗口
// 时间复杂度：O(n)
// 空间复杂度：O(1)
class Solution {
    public int longestOnes(int[] nums, int k) {
        int ans = 0;
        for (int left = 0, right = 0, sum = 0; right < nums.length; ++right) {
            sum += nums[right] == 0 ? 1 : 0;
            if (sum > k) {
                sum -= nums[left++] == 0 ? 1 : 0;
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}

File: leecode/leetcode-cn/1004.最大连续-1-的个数-iii.cpp
/*
 * @lc app=leetcode.cn id=1004 lang=cpp
 *
 * [1004] 最大连续1的个数 III
 */

// @lc code=start
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size(), ans = 0, cnt = 0;  // cnt 统计 0 的个数
        for (int l = 0, r = 0; r < n; ++r) {
            cnt += nums[r] == 0;
            while (cnt > k)
                cnt -= nums[l++] == 0;
            ans = max(ans, r - l + 1);
        }
        return ans;
    }
};
// @lc code=end
